### PR TITLE
Use a merge library to merge cluster- and dc-level specs

### LINKED ISSUE #525: K8SSAND-1460 ⁃ Cluster- and dc-level templates should be deep-merged
**What happened?**

Consider the following `K8ssandraCluster`:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  name: test
spec:
  # Cluster level
  stargate:
    size: 1
    heapSize: 512Mi
  cassandra:
    datacenters:
      - metadata:
          name: dc1
        size: 1
        # DC level
        stargate:
          size: 1
          racks:
            # Rack level
            - name: rack1
```

If I deploy this CRD, the observed heap size for Stargate nodes is `256Mi` (the default).

**Did you expect to see something different?**

I would expect the heap size of all Stargate nodes to be `512Mi`.

This is because the `Coalesce` methods do not merge fields, they simply pick the first non-nil template. In the above example, the Rack template wins over the DC template and the cluster template, so it gets applied, and its heap size is 256Mi (the default).

Note that many e2e are wrongly conceived because of that. They try to set the heap size cluster-wide, yet it is not applied because there is a DC or Rack template somewhere.

**How to reproduce it (as minimally and precisely as possible)**:

Use the above CRD.

**Environment**

* K8ssandra Operator version: 1.0.0

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1460) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1460
┆priority: Medium

### LINKED ISSUE #715: K8SSAND-1821 ⁃ Ability to merge cluster- and dc- level configs
**What is missing?**

Currently cluster-level configs are coalesced with dc-level ones in an all-or-nothing fashion: if the dc-level config is present, then the cluster-level one is ignored.

**Why do we need it?**

With #710, YAML config files are now schemaless in the CRD and represented internally as `map[string]interface{}`. It should be fairly simple to use `utils.MergeMapNested` to merge cluster-level and dc-level configs.

Note: JVM options are still schemaful so we can't apply this fix to the `JvmOptions` struct. We might be able to write custom code to handle it, or resort to merging libraries as suggested in #526.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1821) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1821
┆priority: Medium

### PR DESCRIPTION
**Pull request by adutra**
**What this PR does**:

This PR uses Goalesce to do all the cluster/DC spec merges in the operator code.

**Which issue(s) this PR fixes**:
Fixes #525 
Fixes #715 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: apis/k8ssandra/v1alpha1/k8ssandracluster_types.go [hunk 1]
```diff
@@ -411,6 +411,27 @@ type DatacenterOptions struct {
 	ManagementApiAuth *cassdcapi.ManagementApiAuthConfig `json:"managementApiAuth,omitempty"`
 }
 
+// NetworkingConfig is a copy of cass-operator's NetworkingConfig struct. It is copied here to
+// change the HostNetwork field type from bool to *bool, which makes merging 2 values of this struct
+// more intuitive.
+type NetworkingConfig struct {
```
Comment (adutra) [inline apis/k8ssandra/v1alpha1/k8ssandracluster_types.go:417]: This was a bit disappointing, but there is no way we can merge a `bool` field and expect trilean semantic, so I had to copy the struct here and change to `*bool`.

File: apis/stargate/v1alpha1/stargate_types.go [hunk 2]
```diff
@@ -151,16 +152,20 @@ type StargateDatacenterTemplate struct {
 	Racks []StargateRackTemplate `json:"racks,omitempty"`
 }
 
-// Coalesce compares this StargateDatacenterTemplate with the given StargateClusterTemplate and returns the first
-// non-nil StargateDatacenterTemplate it finds.
-// TODO revisit the merging strategy and/or find a better name for this method
-func (in *StargateDatacenterTemplate) Coalesce(clusterTemplate *StargateClusterTemplate) *StargateDatacenterTemplate {
+// MergeWith merges the given StargateClusterTemplate with this StargateDatacenterTemplate and
+// returns the result.
+func (in *StargateDatacenterTemplate) MergeWith(clusterTemplate *StargateClusterTemplate) *StargateDatacenterTemplate {
```
Comment (adutra) [inline apis/stargate/v1alpha1/stargate_types.go:157]: I had to standardize method names and the general contract. I went for `MergeWith` and the contract is that the receiver is the DC level object, and the passed object is the cluster level one.

File: apis/telemetry/v1alpha1/telemetry_methods.go [hunk 3]
```diff
@@ -1,49 +1,12 @@
...
-		}
-	}
-	if b.CommonLabels != nil {
-		for k, v := range b.CommonLabels {
-			out.CommonLabels[k] = v
-		}
-	}
-	out.Enabled = b.Enabled
-	return &out
+func (in *TelemetrySpec) IsPrometheusEnabled() bool {
```
Comment (adutra) [inline apis/telemetry/v1alpha1/telemetry_methods.go:10]: This isn't  really related to merges but it simplifies the code a lot.

File: apis/telemetry/v1alpha1/telemetry_types.go [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
 
 type TelemetrySpec struct {
 	Prometheus *PrometheusTelemetrySpec `json:"prometheus,omitempty"`
 	Mcac       *McacTelemetrySpec       `json:"mcac,omitempty"`
 }
+
 type PrometheusTelemetrySpec struct {
 	// Enable the creation of Prometheus serviceMonitors for this resource (Cassandra or Stargate).
-	Enabled bool `json:"enabled,omitempty"` // A bool flag required here to disambiguate when e.g. the cluster should have telemetry turned on but one DC should have it explicitly turned off.
+	Enabled *bool `json:"enabled,omitempty"`
```
Comment (adutra) [inline apis/telemetry/v1alpha1/telemetry_types.go:12]: Change required for trilean logic. This has no impact on the CRD definition.

File: apis/k8ssandra/v1alpha1/k8ssandracluster_types.go (hunk 1 above)
Comment (adutra) [reply apis/k8ssandra/v1alpha1/k8ssandracluster_types.go:417]: Note that this has no impact on the CRD definition.

File: docs/arch-overview.md [hunk 5]
```diff
@@ -36,15 +36,15 @@ cassandra:
       size: 3  
 ``` 
 
-There are two elements in the array. K8ssnandra Operator will create CassandraDatacenters named `dc1` and `dc2`. Each one will have a size of 3.
+There are two elements in the array. K8ssandra Operator will create CassandraDatacenters named `dc1` and `dc2`. Each one will have a size of 3.
```
Comment (adutra) [inline docs/arch-overview.md:39]: I didn't change the contents of this file, just fixed the typos.

File: pkg/goalesce/utils.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package goalesceutils
+
+import (
+	"github.com/adutra/goalesce"
```
Comment (adejanovski) [inline pkg/goalesce/utils.go:4]: ahem... can we fork the goalesce project into the k8ssandra org and reference that version instead?

Comment (adutra) [reply pkg/goalesce/utils.go:4]: As we discussed offline, this is a project that I started outside K8ssandra, and I intend to maintain it. Of course you could fork it if you want to, but at this point, I don't see any valid reason to do so.

Comment (adejanovski) [reply pkg/goalesce/utils.go:4]: sounds good 👍

**Review by adejanovski — APPROVED** [review]

**Pull request merged**